### PR TITLE
Implement removing leading whitespace from un-evaluated String expressions

### LINKED ISSUE #4854: Handle leading whitespace in WDL command blocks more correctly
[The WDL spec](https://github.com/openwdl/wdl/blob/main/versions/1.1/SPEC.md#stripping-leading-whitespace) says we need to remove "common leading whitespace". For some reason, as noted in https://github.com/DataBiosphere/wdl-conformance-tests/pull/34, we're apparently removing *all* whitespace equal to the *first* indentation level on any line it appears on. If lines dedent out from the leading line, we probably need to not remove *any* whitespace that isn't common to *all* the lines.

I think we might be doing this to work around https://github.com/chanzuckerberg/miniwdl/issues/674, but we also might have come up with a better workaround now and can maybe change the behavior?

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1531)
┆Issue Number: TOIL-1531

### PR DESCRIPTION
**Pull request by adamnovak**
This should fix #4854 by implementing the spec behavior of removing common leading whitespace *before* substituting in the values in the expressions.

It still can't handle a line that begins with a placeholder or an expression that has whitespace *within* placeholders; it will find the common leading whitespace of other lines and remove that.

This will break some workflows that expect to be able to dedent beyond where the command string starts. I have an argument to the function for this but it's not bound out to the command line.

This should have tests, but currently doesn't, beyond the conformance tests.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * `toil-wdl-runner` now removes whitespace *before* evaluating command strings.
 
## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

File: src/toil/test/wdl/wdltoil_test_kubernetes.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def test_wdl_kubernetes_cluster(self):
         self.leader = self.cluster.getLeader()
 
         url = "https://github.com/DataBiosphere/wdl-conformance-tests.git"
-        commit = "09b9659cd01473e836738a2e0dd205df0adb49c5"
+        commit = "7c7f82c0dd426f5be708f7a858e448cfb0fea91b"
```
Comment (DailyDreaming) [inline src/toil/test/wdl/wdltoil_test_kubernetes.py:54]: Why is this commit hash different from the one in `src/toil/test/wdl/wdltoil_test.py`?

File: src/toil/wdl/wdltoil.py [hunk 2]
```diff
@@ -142,6 +142,184 @@ def decorated(*args: Any, **kwargs: Any) -> Any:
...
+        the common whitespace.
+
+    :param tolerate_dedents: If True, remove as much of the whitespace on the
+        first indented line as is found on subesquent lines, regardless of
+        whether later lines are out-dented relative to it.
+
+    :param tolerate_all_whitespace: If True, don't allow all-whitespace lines
+        to reduce the common whitespace prefix.
+
+    :param debug: If True, log intermediate staged of processing.
```
Comment (DailyDreaming) [inline src/toil/wdl/wdltoil.py:163]: Could this be phrased more clearly?

File: src/toil/wdl/wdltoil.py [hunk 3]
```diff
@@ -142,6 +142,184 @@ def decorated(*args: Any, **kwargs: Any) -> Any:
         return cast(F, decorated)
     return decorator
 
+def remove_common_leading_whitespace(expression: WDL.Expr.String, tolerate_blanks: bool = True, tolerate_dedents: bool = False, tolerate_all_whitespace: bool = True, debug: bool = False) -> WDL.Expr.String:
```
Comment (DailyDreaming) [inline src/toil/wdl/wdltoil.py:145]: Could there be a test for this function?

**Review by DailyDreaming — COMMENTED** [review]
LGTM.  Just a couple of questions and, due to the complexity of the function, would like a test (something that ideally shows what the exact expected behavior should be for the function and enforces maintaining that).

File: src/toil/wdl/wdltoil.py (hunk 3 above)
Comment (adamnovak) [reply src/toil/wdl/wdltoil.py:145]: I've added some tests.

File: src/toil/wdl/wdltoil.py (hunk 2 above)
Comment (adamnovak) [reply src/toil/wdl/wdltoil.py]: I've updated the parameter description.

File: src/toil/test/wdl/wdltoil_test_kubernetes.py (hunk 1 above)
Comment (adamnovak) [reply src/toil/test/wdl/wdltoil_test_kubernetes.py]: I've changed the commit for both tests to come from just one place.

**Review by DailyDreaming — APPROVED** [review]
Looks great.  Thanks for making the changes.

**Pull request merged**